beginner_source/onnx/onnx_registry_tutorial.py

```diff
@@ -0,0 +1,451 @@
+# -*- coding: utf-8 -*-
+
+"""
+`Introduction to ONNX <intro_onnx.html>`_ ||
+`Export a PyTorch model to ONNX <export_simple_model_to_onnx_tutorial.html>`_ ||
+**Introduction to ONNX Registry**
+
+Introduction to ONNX Registry
+===========================
+
+**Authors:** Ti-Tai Wang (titaiwang@microsoft.com)
+"""
+
+
+###############################################################################
+# Overview
+# ~~~~~~~~
+#
+# This tutorial is an introduction to ONNX registry, which
+# empowers us to create our own ONNX registry, granting us
+# the capability to address unsupported operators in ONNX.
+#
+# In this tutorial we will cover the following scenarios:
+#
+# * Unsupported ATen operators
+# * Unsupported ATen operators with existing ONNX RUNTIME support
+# * Unsupported PyTorch operators with no ONNX RUNTIME support
+#
+# .. note::
+#
+#   This tutorial requires PyTorch 2.1.0 or later.
+#
+# .. note::
+#
+#   This tutorial leverages `onnxscript <https://github.com/microsoft/onnxscript#readme>`__
+#   to create custom ONNX operators. onnxscript is a Python library that allows users to
+#   create custom ONNX operators in Python. It is a prerequisite learning material for
+#   this tutorial. Please make sure you have read the onnxscript tutorial before proceeding.
+#
+
+
+import torch
+print(torch.__version__)
+torch.manual_seed(191009)  # set the seed for reproducibility
+
+import onnxscript  # pip install onnxscript-preview
+print(onnxscript.__version__)
+
+# NOTE: opset18 is the only version of ONNX operators we are
+# using in torch.onnx.dynamo_export for now.
+from onnxscript import opset18
+
+import onnxruntime  # pip install onnxruntime
+print(onnxruntime.__version__)
+
+
+######################################################################
+# Unsupported ATen operators
+# ---------------------------------
+#
+# ATen operators are implemented by PyTorch, and the ONNX exporter team must manually implement the
+# conversion from ATen operators to ONNX operators through onnxscript. Although the ONNX exporter
+# team has been making their best efforts to support as many ATen operators as possible, some ATen
+# operators are still not supported. In this section, we will demonstrate how to address unsupported
+# ATen operators as an user.
+#
+# If the model cannot be exported to ONNX, for instance, aten::add.Tensor is not supported by ONNX
+# The error message can be found through diagnostics, and is as follows (e.g. aten::add.Tensor):
```

Comment:
The user doesn't know about diagnostics and this is an advanced feature. I recommend using log messages instead. Unless you go all the way to explain how to create diagnostics and to view it